text/3892-complex-numbers.md

```diff
@@ -0,0 +1,252 @@
+
+## Motivation
+[motivation]: #motivation
+
+The definition of complex numbers in the C99 standard defines the _memory layout_ of a complex number but not its _calling convention_. 
+This makes crates like `num-complex` untenable for calling C FFI functions containing complex numbers without at least a level of indirection (`*const Complex`) or the like.
+Only in `std` is it possible to make an additional repr to match the calling convention that C uses across FFI boundaries. 
+In essence, this RFC makes code like this:
+```C
+extern double _Complex computes_function(double _Complex x);
+```
+callable in Rust without indirection:
+```rust
+extern "C" {
+  fn computes_function(x: Complex<f64>) -> Complex<f64>;
+}
+fn main() {
+  let returned_value = computes_function(Complex<f64>::new(3, 4))
+}
+```
+using the standard library's FFI-compatible complex numbers.
+
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`Complex<T>` numbers can be instantiated as of any type using `Complex::new(re, im)` where `re` and `im` are of the same type (this includes all numbers).
+```rust
+let x = Complex::new(3.0, 4.0);
+```
+They can even be passed as an array:
+```rust
+let y = Complex::from([3.0, 4.0]);
+```
+or as a tuple:
+```rust
+let z = Complex::from((3.0, 4.0));
+```
+They can even be passed in polar form (but only as a float):
+```rust
+let polar = Complex::from_polar(3.0, f32::PI/2.0);
+```
+They are added and multiplied as complexes are:
+```rust
+let first = Complex::new(1.0, 2.0);
+let second = Complex::new(3.0, 4.0);
+let added = first + second; // 4 + 6i
+let multiplied = first * second; // -4 + 10i
+```
+
+They can be divided using normal floating-point division:
+```rust
+let float_first = Complex::new(1.0, 2.0);
+let float_second = Complex::new(3.0, 4.0);
+let divided = float_second / float_first; // 2.4 - 0.2i
+```
+
+You can even calculate the complex sine, cosine and more:
+```rust
+let val = Complex::new(3.0, 4.0);
+let sine_cmplx = csin(val); // 3.8537380379 - 27.016813258i
+```
+If you want to call certain C libraries with complex numbers, you use this type:
+```C
+// in the C library
+extern double _Complex computes_function(double _Complex x);
+```
+```rust
+// in YOUR Rust code
+extern "C" {
+  fn computes_function(x: Complex<f64>) -> Complex<f64>;
+}
+fn main() {
+  let returned_value = computes_function(Complex::<f64>::new(3.0, 4.0))
+}
+```
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Complex numbers will be implemented by using traits in the `core` crate:
+```
+trait Float: Copy + Clone {}
+impl Float for f32 {}
+impl Float for f64 {}
+```
+Calls to some `libgcc` functions will also be needed and will be emitted by the backend via compiler-builtins, specifically `__mulsc3`, `__muldc3`, `__divsc3` and `__divdc3` for the proper and complete implementation of these types.
```

Comment:
Presumably these impls will be in `std`, and `core` will have the `Complex` type but without a full set of arithmetic impls, so `Complex` can still be used without dependency on system libraries. I think that might be worth describing more precisely and explicitly, for readers concerned with `no_std` use cases.